stablehlo/conversions/tosa/transforms/StablehloQuantLegalizeToTosaRescale.cpp

```diff
@@ -0,0 +1,257 @@
+  void runOnOperation() final;
+
+  LogicalResult initialize(MLIRContext* ctx) override;
+
+ private:
+  FrozenRewritePatternSet patterns;
+};
+
+#define DECL_CONVERT_OP(stablehlo_op)                                        \
+  struct ConvertStablehloQuant##stablehlo_op##Op                             \
+      : public OpRewritePattern<stablehlo::stablehlo_op##Op> {               \
+    using OpRewritePattern<stablehlo::stablehlo_op##Op>::OpRewritePattern;   \
+    LogicalResult matchAndRewrite(stablehlo::stablehlo_op##Op op,            \
+                                  PatternRewriter& rewriter) const override; \
+  }
+
+DECL_CONVERT_OP(Add);
+DECL_CONVERT_OP(Subtract);
+
+#undef DECL_CONVERT_OP
+
+// create a tosa rescale op and return its result value
+Value buildRescale(PatternRewriter& rewriter, Location loc,
+                   ShapedType output_type, Value input_val, int32_t multiplier,
+                   int32_t shift, int64_t inputZp, int64_t outputZp,
+                   bool double_round, bool scale32, bool perChannel) {
+  // bool input_unsigned = input_val.getType().isUnsignedInteger();
+  // bool output_unsigned = output_type.isUnsignedInteger();
+
+  auto rescale_op = rewriter.create<tosa::RescaleOp>(
+      loc, output_type, input_val,
+      rewriter.getI32IntegerAttr(static_cast<int32_t>(inputZp)),
+      rewriter.getI32IntegerAttr(static_cast<int32_t>(outputZp)),
+      rewriter.getDenseI32ArrayAttr({multiplier}),
+      rewriter.getDenseI8ArrayAttr({static_cast<int8_t>(shift)}),
+      rewriter.getBoolAttr(scale32), rewriter.getBoolAttr(double_round),
+      rewriter.getBoolAttr(perChannel));
+
+  return rescale_op.getResult();
+}
+
+// Creates TOSA rescale op with int32 output
+Value buildRescaleToInt32(PatternRewriter& rewriter, Location loc,
+                          Value inputVal, double inputScale, int64_t inputZp) {
+  auto inputType = dyn_cast<mlir::ShapedType>(inputVal.getType());
+  if (!inputType) {
+    return nullptr;
+  }
+  auto outputType = inputType.clone(rewriter.getI32Type());
+
+  const int32_t scaleWidth = 32;
+  int32_t multiplier, shift;
+  computeMultiplierAndShift(inputScale, multiplier, shift, scaleWidth);
+
+  return buildRescale(rewriter, loc, outputType, inputVal, multiplier, shift,
+                      inputZp,
+                      /* output_zp = */ 0,
+                      /* double_round = */ false,
+                      /* scale32 = */ true,
+                      /* per_channel = */ false);
+}
+
+// Creates TOSA rescale op with int32 input
+Value buildRescaleFromInt32(PatternRewriter& rewriter, Location loc,
+                            mlir::ShapedType outputType, Value inputVal,
+                            double outputScale, int64_t outputZp) {
+  // Input should be int32 type
+  auto inputType = dyn_cast<mlir::ShapedType>(inputVal.getType());
+  if (!inputType || !inputType.getElementType().isInteger(32)) {
+    // expected rescale input element type to be i32
```

Comment:
remove comment?

Reply:
Use assert similar to https://github.com/tensorflow/tensorflow/blob/79bac912c5fee54986b79fef8eb7aabdfa254560/tensorflow/compiler/mlir/tosa/transforms/legalize_utils.cc#L209

Reply:
done